pgml-extension/src/proxy.rs

```diff
@@ -0,0 +1,74 @@
+//! Proxy API calls to a Postgres BG worker
+use std::{
+    io::{self, Read, Write},
+    net::{SocketAddr, TcpStream},
+    time::Duration,
+};
+
+use once_cell::sync::OnceCell;
+use pgrx::{bgworkers::BackgroundWorkerBuilder, error, IntoDatum};
+use proxy_message::{Reply, Request, REPLY_SIZE};
+
+use crate::config::get_config;
+
+const RESTART_DURATION: Duration = Duration::from_secs(1);
+static NAME: &str = "pgml_proxy";
+static PROXY_ADDR: OnceCell<SocketAddr> = OnceCell::new();
+static CONFIG_LIBRARY_KEY: &str = "pgml.proxy.library";
+static CONFIG_FUNCTION_KEY: &str = "pgml.proxy.entry_fn";
+static CONFIG_SOCKET_ADDR_KEY: &str = "pgml.proxy.socket_addr";
+
+macro_rules! get_config_or_return {
+    ($name:expr) => {
+        match get_config($name) {
+            Some(v) => v,
+            None => return,
+        }
+    };
+}
+
+/// Setup the proxy bg_worker using values in `postgresql.conf`.
+pub fn setup() {
+    let library = get_config_or_return!(CONFIG_LIBRARY_KEY);
+    let function = get_config_or_return!(CONFIG_FUNCTION_KEY);
+    let socket_addr = get_config_or_return!(CONFIG_SOCKET_ADDR_KEY);
+
+    load_bg_worker(&library, &function, &socket_addr);
+}
+
+/// Returns whether a proxy is being used.
+pub fn enabled() -> bool {
+    PROXY_ADDR.get().is_some()
+}
+
+fn load_bg_worker(library: &str, function: &str, socket_addr: &str) {
+    match socket_addr.parse::<SocketAddr>() {
+        Ok(addr) => {
+            // UNWRAP: This is the only place it is set and should only be called once
+            PROXY_ADDR.set(addr).unwrap();
+        }
+        Err(e) => error!("could not parse proxy address: {e}"),
+    };
+
+    BackgroundWorkerBuilder::new(NAME)
+        .set_function(function)
+        .enable_shmem_access(None)
+        .set_library(library)
+        .set_restart_time(Some(RESTART_DURATION))
+        .set_argument(socket_addr.into_datum())
+        .load();
+}
+
+/// Send a `Request` to the proxy and wait for a `Reply`.
+pub fn send_request(request: Request) -> io::Result<Reply> {
+    let addr = PROXY_ADDR.get().unwrap();
+    let mut stream = TcpStream::connect(addr)?;
```

Comment:
You want to keep a connection open, otherwise this will be pretty slow (3 roundtrips to establish a connection).

Also I think you want a UNIX socket here.

Reply:
Yeah I don't know what I was thinking.

Reply:
https://github.com/HyperparamAI/zoomies/blob/master/src/lib.rs has a datagram example over udp for unix domain socket.

Reply:
If we keep a connection open and reuse it, then we won't get parallel use of the proxy service because we are serializing through the single shared connection. Do we need a connection pool?

Reply:
We don't need a pool in the sense I'm familiar with, but we do need an independent response socket (or some other individually addressed mechanism) for each process that is sending messages on the primary bg_worker socket that also acts as a work queue, so client foreground processes can wait for their individual responses.

i.e. every process sends their request, along with a response addressing mechanism (their process id) to the bg_worker on a shared socket with multiple datagram writers (connection processes) and a single datagram reader (background worker). The bg_worker then connects to a new response socket per process, where only the sender is listening the first time it gets a request from the foreground process.

Reply:
The client (pgml) should maintain an open connection. Each new Postgres connection process will open its own connection. The server (bg worker) will listen on the socket and allocate a new connection for each new client.

Reply:
> The client (pgml) should maintain an open connection. Each new Postgres connection process will open its own connection.

I'm not sure I understand this point. But hopefully the new commits will help me understand how far off I am from this thinking.